Report/Chapters/Design.tex

```diff
@@ -8,5 +8,50 @@ \section{Languages and libraries}
 
 \section{High-level architecture}
 
-% Frontend/backend split
-% Communication between frontend and backend (when we've got that settled)
+Viskell consists of two parts, a front end and a back end.
+The front end is responsible for providing a user interface, handling user input, displaying error messages, and communicating with the back end.
+The back end consists of a representation for Haskell code in Java, an interface to GHCi and functionality to make working with the back end easier.
+This separation is made to make development and debugging for the part that generates Haskell code easier.
+It also makes using the user interface for a different purpose a bit easier, although the user interface is not specifically designed for this purpose.
+
+\section{Backend architecture}
+
+A result of our design choices is that our Java program needs to have a basic understanding of the Haskell programming
+language. For this purpose we implemented two tree structures - one for expressions, one for types - supported by an
+interface for GHCi, a type signature parser and a catalog.
+
+\subsection{Communication with GHCi}
+\index{GHCi}
+
+The base of the back-end is an interface for communication with GHCi. This interface allows us to run Haskell code
+generated by the users of our application. As implementing a complete type checker for Haskell is a major challenge this
+interface can also be used to check a program for faults.
```

Comment:
As implementing a complete type checker for Haskell is a major challenge this interface can also be used to check a program for faults.

I don't understand what you are trying to say here.